### PR TITLE
Use fast property setter in MutableObjectModelBinder

### PR DESCRIPTION
**Pull request by ajaybhargavb**
Issue - #2259

### CODE REVIEW

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 1]
```diff
@@ -286,7 +286,7 @@ public override bool IsReadOnly
                     }
                     else
                     {
-                        _isReadOnly = _details.PropertySetter != null;
+                        _isReadOnly = _details.PropertySetter == null;
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: Yep, looks like we had both of these checks backwards.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 2]
```diff
@@ -407,5 +407,23 @@ public override IReadOnlyList<object> ValidatorMetadata
...
+            }
+        }
+
+        /// <inheritdoc />
+        public override Action<object, object> PropertySetter
+        {
+            get
+            {
+                return _details.PropertySetter;
+            }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: Would like to see some unit tests for when these are populated and not. Take a look at `ModelMetadataProviderTest` which has more of the integration type tests.

**Comment by rynowak** [comment]
:watch: for tests

File: src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs [hunk 3]
```diff
@@ -480,7 +480,7 @@ internal void ProcessDto(ModelBindingContext bindingContext, ComplexModelDto dto
             {
                 try
                 {
-                    property.SetValue(bindingContext.Model, value);
+                    propertyMetadata.PropertySetter(bindingContext.Model, value);
```
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: There's code on Line 430 that does reflection to fetch all properties. Could you file a follow up work item to remove using reflection and use model metadata?

Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Binders/MutableObjectModelBinder.cs]: This is already tracked (assigned me to me)

**Comment by ajaybhargavb** [comment]
Updated.

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs [hunk 4]
```diff
@@ -284,9 +284,13 @@ public override bool IsReadOnly
                     {
                         _isReadOnly = BindingMetadata.IsReadOnly;
                     }
+                    else if (MetadataKind == ModelMetadataKind.Property)
+                    {
+                        _isReadOnly = _details.PropertySetter == null;
+                    }
```
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadata.cs]: This block should probably come first (ignore `BindingMetadata.IsReadOnly` if it's Type metadata).

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs [hunk 5]
```diff
@@ -411,8 +411,72 @@ public void PropertiesEnumerationEvaluatedOnce()
             Assert.Equal(firstPropertiesEvaluation, secondPropertiesEvaluation);
         }
 
+        [Fact]
+        public void IsReadOnly_returnsFalse_ForType()
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: `returnsFalse`

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 6]
```diff
@@ -696,11 +696,53 @@ public void IsRequired_DataMemberIsRequiredTrueWithoutDataContract_False()
...
+            // Act
+            var metadata = metadataProvider.GetMetadataForProperty(
+                typeof(ClassWithPrivateSetProperty),
+                nameof(ClassWithPrivateSetProperty.StringProperty));
+
+            // Assert
+            Assert.Null(metadata.PropertySetter);
+            Assert.NotNull(metadata.PropertyAccessor);
+        }
+
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: How about the getter?

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: Added a test.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs [hunk 7]
```diff
@@ -696,11 +696,75 @@ public void IsRequired_DataMemberIsRequiredTrueWithoutDataContract_False()
...
+            var type = typeof(ClassWithPrivateGetProperty);
+            var propertyName = nameof(ClassWithPrivateGetProperty.StringProperty);
+
+            // Act & Assert
+            var exception = Assert.Throws<ArgumentException>(() =>
+            {
+                metadataProvider.GetMetadataForProperty(type, propertyName);
+            });
+            Assert.Equal($"The property {type}.{propertyName} could not be found.{Environment.NewLine}Parameter name: propertyName",
+                exception.Message);
```
Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataProviderTest.cs]: Just do `GetMetadataForType` and then check that the property is null, will be clearer.

**Comment by rynowak** [comment]
:shipit:

**Comment by pranavkm** [comment]
:shipit:

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs [hunk 8]
```diff
@@ -411,8 +411,72 @@ public void PropertiesEnumerationEvaluatedOnce()
...
+            Assert.False(isReadOnly);
+        }
+
         private void ActionMethod(string input)
         {
         }
+
+        private class Foo
+        {
+            public string Bar { private set; get; }
```
Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: Recommend renaming this to be something like PublicGetPrivateSetProperty and PublicGetPublicSetProperty to make the rest of the code easier to read.

BTW are scenarios like protected properties and/or protected getters/setters interesting here?

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs [hunk 9]
```diff
@@ -289,5 +289,15 @@ public string GetDisplayName()
         {
             return DisplayName ?? PropertyName ?? ModelType.Name;
         }
+
+        /// <summary>
+        /// Gets or sets a property accessor delegate to get the property value from a model object.
+        /// </summary>
+        public abstract Func<object, object> PropertyAccessor { get; }
```
Comment (Eilon) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs]: Rename to `PropertyGetter`? (Matches `PropertySetter` better.)

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs (hunk 8 above)
Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: I don't think so. The condition is for them to not be private. So it doesn't matter if they are protected or public.

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: You mean the condition is for them *to be public* right? We presumably want to ignore protected/private/internal?

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: Right now the PropertySetter is being set for anything that is not private set (similarly for the getter). Should we change it to allow only public? https://github.com/aspnet/Mvc/pull/2342/files#diff-8a66a6db79787dca4cf9119d59910ec1R128

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: Hmm that looks suspicious to me. @rynowak @harshgMSFT any thoughts on what the check should be?

Comment (rynowak) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: The right behavior is whatever MVC5 did. As for the reflection APIs, they don't always give top billing to the other access modifiers than public or private. I seems to recall a few cases where private means non-public.

Comment (Eilon) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: Agreed.

Comment (ajaybhargavb) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: Checked MVC 5's behavior. Anything not public is considered ReadOnly. Will update the condition to reflect the same.

**Comment by ajaybhargavb** [comment]
Updated @Eilon

**Comment by Eilon** [comment]
Looks good! :shipit:

File: src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs [hunk 10]
```diff
@@ -53,7 +53,7 @@ public DefaultMetadataDetails(ModelMetadataIdentity key, IReadOnlyList<object> a
         /// <summary>
         /// Gets or sets a property accessor delegate to get the property value from a model object.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultMetadataDetails.cs]: property getter delegate

File: src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs [hunk 11]
```diff
@@ -299,5 +299,15 @@ public string GetDisplayName()
         {
             return DisplayName ?? PropertyName ?? ModelType.Name;
         }
+
+        /// <summary>
+        /// Gets or sets a property accessor delegate to get the property value from a model object.
```
Comment (harshgMSFT) [inline src/Microsoft.AspNet.Mvc.ModelBinding/ModelMetadata.cs]: property getter delegate

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs [hunk 12]
```diff
@@ -411,8 +411,93 @@ public void PropertiesEnumerationEvaluatedOnce()
...
+            Assert.False(isReadOnly);
+        }
+
         private void ActionMethod(string input)
         {
         }
+
+        private class Foo
+        {
+            public string PublicGetPrivateSetProperty { private set; get; }
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: nit: feels a bit weird looking at order of get and set reversed. Upto you if you want to fix it.

File: test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs [hunk 13]
```diff
@@ -411,8 +411,93 @@ public void PropertiesEnumerationEvaluatedOnce()
...
+
+            // Assert
+            Assert.False(isReadOnly);
+        }
+
         private void ActionMethod(string input)
         {
         }
+
+        private class Foo
```
Comment (harshgMSFT) [inline test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/DefaultModelMetadataTest.cs]: even though this is a private class consider naming it with a better name.

**Comment by harshgMSFT** [comment]
fix comments and optional nits then :shipit:

**Pull request merged**